### PR TITLE
Implement `System::User` on Windows

### PR DESCRIPTION
**Pull request by HertzDevil**
Depends on #14929.

This is for the most part a straight port of [Go's implementation](https://cs.opensource.google/go/go/+/refs/tags/go1.23.0:src/os/user/lookup_windows.go), including their interpretation of primary groups on Windows (as opposed to [whatever Cygwin does](https://cygwin.com/cygwin-ug-net/ntsec.html)).

### CODE REVIEW

File: src/crystal/system/win32/user.cr [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+      begin
+        from_sid(sid)
+      ensure
+        LibC.LocalFree(sid)
+      end
+    end
+  end
+
+  private def self.from_sid(sid : LibC::SID*) : ::System::User?
+    canonical = sid_to_name(sid) || return
```
Comment (Sija) [inline src/crystal/system/win32/user.cr:92]:
```suggestion
    return unless canonical = sid_to_name(sid)
```

File: src/crystal/system/win32/user.cr [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+        str, _ = String.from_utf16(info.value.usri10_full_name)
+        str
+      ensure
+        LibC.NetApiBufferFree(info)
+      end
+    end
+  end
+
+  private def self.lookup_primary_group_id(name : String, domain : String) : String?
+    domain_sid = name_to_sid(domain) || return
```
Comment (Sija) [inline src/crystal/system/win32/user.cr:135]:
```suggestion
    return unless domain_sid = name_to_sid(domain)
```

**Review by Sija — COMMENTED** [review]
Style related suggestions.

**Comment by straight-shoota** [comment]
I noticed the Go implementation has quite some helpful code comments that explain how the code works and why.
Could we have some of that in the Crystal implementation as well?

File: src/crystal/system/win32/user.cr (hunk 1 above)
Comment (Sija) [reply src/crystal/system/win32/user.cr:92]: @straight-shoota Why thumbs down? It's established pattern, used one line above it...

Comment (straight-shoota) [reply src/crystal/system/win32/user.cr:92]: Not this is not an established pattern. And it never will be (if it's up to me).
In a trailing conditional, the condition should always be a predicate without side effects.
Assignments should never be in a trailing condition.

The current style is concise and expressive. It focuses on the main action, an assignment to a local variable. The value that we want to assign comes from a method call and if it is unavailable, we return. That's very readable. Your suggestion is not. It hides the fact that this is an assignment behind the less significant control flow action.

**Review by straight-shoota — APPROVED** [review]

File: src/crystal/system/win32/user.cr (hunk 1 above)
Comment (ysbaddaden) [reply src/crystal/system/win32/user.cr:92]: TBH I write `return unless x = y` in all my crystal code.

Comment (Sija) [reply src/crystal/system/win32/user.cr:92]: @ysbaddaden 💯

**Pull request merged**